modules/4_mnist/src/mnist.cpp

```diff
@@ -22,49 +23,128 @@ void loadImages(const std::string& filepath,
 
     int numImages = readInt(ifs);
 
-    // TODO: follow "FILE FORMATS FOR THE MNIST DATABASE" specification
-    // at http://yann.lecun.com/exdb/mnist/
+	int n_rows = readInt(ifs);
+	int n_cols = readInt(ifs);
+	char* pixels = new char[n_rows * n_cols];
+	for (int item_id = 0; item_id < numImages; ++item_id) {
+		ifs.read(pixels, n_rows * n_cols);
+		cv::Mat image_tmp(n_rows, n_cols, CV_8UC1, pixels);
+		Mat m2 = image_tmp.clone();
+		images.push_back(m2);
+	}
 }
 
 void loadLabels(const std::string& filepath,
-                std::vector<int>& labels) {
-    std::ifstream ifs(filepath.c_str(), std::ios::binary);
-    CV_CheckEQ(ifs.is_open(), true, filepath.c_str());
-
-    int magicNum = readInt(ifs);
-    CV_CheckEQ(magicNum, 2049, "");
-
-    int numLabels = readInt(ifs);
-
-    // TODO: follow "FILE FORMATS FOR THE MNIST DATABASE" specification
-    // at http://yann.lecun.com/exdb/mnist/
+	std::vector<int>& labels) {
+	std::ifstream ifs(filepath.c_str(), std::ios::binary);
+	CV_CheckEQ(ifs.is_open(), true, filepath.c_str());
+
+	int magicNum = readInt(ifs);
+	CV_CheckEQ(magicNum, 2049, "");
+
+	int numLabels = readInt(ifs);
+	char label;
+	// TODO: follow "FILE FORMATS FOR THE MNIST DATABASE" specification
+	// at http://yann.lecun.com/exdb/mnist/
+
+	for (int item_id = 0; item_id < numLabels; ++item_id) {
+		// read label
+		ifs.read(&label, 1);
+		labels.push_back(int(label));
+	}
 }
 
 void prepareSamples(const std::vector<cv::Mat>& images, cv::Mat& samples) {
-    CV_Error(Error::StsNotImplemented, "prepareSamples");
+	
+	for (int i = 0; i < images.size(); i++)
+	{
+		samples.push_back(images[i].reshape(1, 1));
+	}
+	samples.convertTo(samples, CV_32F);
 }
 
+using namespace ml;
+
 Ptr<ml::KNearest> train(const std::vector<cv::Mat>& images,
                         const std::vector<int>& labels) {
-    CV_Error(Error::StsNotImplemented, "train");
+	Ptr<ml::TrainData> trainingData;
+	Ptr<ml::KNearest> kclassifier = KNearest::create();
+	cv::Mat samples;
+	prepareSamples(images, samples);
+	trainingData = TrainData::create(samples,
+		SampleTypes::ROW_SAMPLE, labels);
+	kclassifier->setIsClassifier(true);
+	kclassifier->setAlgorithmType(KNearest::Types::BRUTE_FORCE);
+	kclassifier->setDefaultK(1);
+	kclassifier->train(trainingData);
```

Comment:
Will test for a water counter pass without 
```cpp
kclassifier->setIsClassifier(true);
kclassifier->setAlgorithmType(KNearest::Types::BRUTE_FORCE);
kclassifier->setDefaultK(1);
```
?

Reply:
Yes, thanks!